R/tmb-methods.R

```diff
@@ -429,3 +429,114 @@ h_residuals_response <- function(object) {
+
+  method <- match.arg(method)
+
+  if(is.null(newdata)){
+    # use data in fit
+    newdata <- object$data
+  }
+
+  # process the new data such that new data has same levels as original data
+  full_frame <- model.frame(
+    object,
+    data = newdata,
+    include = c("subject_var", "visit_var", "group_var", "response_var"),
+    na.action = "na.pass"
+  )
+  tmb_data <- h_mmrm_tmb_data(
+    object$formula_parts, full_frame,
+    weights = rep(1, nrow(full_frame)),
+    reml = TRUE,
+    singular = "keep",
+    drop_visit_levels = FALSE,
+    allow_na_response = TRUE,
+    drop_levels = FALSE
+  )
+
+  if(method == "conditional"){
+    # get prediction and prediction variance
+    mu <- h_get_prediction(tmb_data, object$theta_est, object$beta_est, object$beta_vcov)
+
+    # build data.frame of prediction vectors
+    ret <- mu$prediction[,1]
+    ret <- data.frame(replicate(nsim, ret))
+
+    # get dataframe values
+    n_subjects <- tmb_data$n_subjects
+    subject_n_visits <- tmb_data$subject_n_visits
+    subject_zero_inds <- tmb_data$subject_zero_inds
+    visit_zero_inds <- tmb_data$visits_zero_inds
+
+      for(i in seq_len(n_subjects)){
+        if(length(mu$index[[i]]) > 0){
+          # Obtain indices of data.frame belonging to subject i (iterate by 1, since indices from cpp are 0-order)
+          inds <- mu$index[[i]] + 1
+
+          # get relevant covariance matrix for subject i
+          covmat_i <- mu$covariance[[i]]
+
+          # simulate from covariance matrix
+          ret[inds,] <- ret[inds, , drop=FALSE] + t(MASS::mvrnorm(nsim, rep.int(0, length(inds)), covmat_i))
+        }
+      }
+  }else if(method == "marginal"){
+    # build data.frame of prediction vectors
+    ret <- as.data.frame(matrix(NA, ncol = nsim, nrow = nrow(newdata)))
+
+    # get dataframe values
+    n_subjects <- tmb_data$n_subjects
+    subject_n_visits <- tmb_data$subject_n_visits
+    subject_zero_inds <- tmb_data$subject_zero_inds
+    visit_zero_inds <- tmb_data$visits_zero_inds
+
+    for(j in seq_len(nsim)){
+      # get prediction and prediction variance
+
+      # sample from theta distribution
+      newtheta <- MASS::mvrnorm(1, object$theta_est, object$theta_vcov)
+      # recalculate betas with sampled thetas
+      hold <- object$tmb_object$report(newtheta)
+      newbeta <- hold$beta
+      newbeta_vcov <- hold$beta_vcov
+
+      # get new predictions
+      mu <- h_get_prediction(tmb_data, newtheta, newbeta, newbeta_vcov)
+      ret[,j] <- mu$prediction[,1]
+
+      for(i in seq_len(n_subjects)){
+        if(length(mu$index[[i]]) > 0){
+          # Obtain indices of data.frame belonging to subject i
+          inds <- mu$index[[i]] + 1
+
+          # get relevant covariance matrix for subject i
+          covmat_i <- mu$covariance[[i]]
+
+          # simulate from covariance matrix
+          ret[inds,j] <- ret[inds, j, drop=FALSE] + t(MASS::mvrnorm(1, rep.int(0, length(inds)), covmat_i))
+          }
+        }
```

Comment:
this part seems a bit duplicated with the previous part; is it possible to reduce the duplication, like

```
h_fun <- function(mu, cov, n_sim) {
xx
}
```
so that in the conditional part we have

```
h_fun(mu, cov, nsim)
```

and in marginal part we have

```
lapply(seq_len(nsim), function(x) {
derivation of new theta; mu; cov; 
h_fun(mu, cov, 1L)
}
```